### PR TITLE
Version 3 - generator of templates

### PR DESCRIPTION
**Pull request by Truba**
### Version 3
* A lot more initial templates
* Creating templates through a ruby script
* RxSwift support
* Community oriented
* Simple, robust and minimalistic interfaces

### CODE REVIEW

File: Templates/Resources/Templates/___FILEBASENAME___Interfaces.swift.erb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+protocol ___VARIABLE_moduleName___PresenterInterface: PresenterInterface {
+<% if @template.generate_io -%>
+    func configure(with output: ___VARIABLE_moduleName___.ViewOutput) -> ___VARIABLE_moduleName___.ViewInput
+<% end -%>
+}
+
+<% if @complexity.generate_formatter -%>
+protocol ___VARIABLE_moduleName___FormatterInterface: FormatterInterface {
+    <%- if @template.generate_io -%>
+    func format(for input: ___VARIABLE_moduleName__.FormatterInput) -> ___VARIABLE_moduleName__.FormatterOutput
```
Comment (fgulan) [inline Templates/Resources/Templates/___FILEBASENAME___Interfaces.swift.erb]: Missing one `_` at the end of both `___VARIABLE_moduleName__`

File: Templates/Resources/Templates/___FILEBASENAME___ViewController.swift.erb [hunk 2]
```diff
@@ -20,6 +24,9 @@ final class ___VARIABLE_moduleName___ViewController: UIViewController {
 
     override func viewDidLoad() {
         super.viewDidLoad()
+        <%- if @template.generate_io -%>
+        configure()
```
Comment (fgulan) [inline Templates/Resources/Templates/___FILEBASENAME___ViewController.swift.erb]: Private underscore

File: Templates/Resources/Templates/___FILEBASENAME___Formatter.swift.erb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+<% end -%>
+
+final class ___VARIABLE_moduleName___Formatter {
+}
+
+// MARK: - Extensions -
+
+extension ___VARIABLE_moduleName___Formatter: ___VARIABLE_moduleName___FormatterInterface {
+    <%- if @template.generate_io -%>
+    func format(for input: ___VARIABLE_moduleName___.ForlatterInput) -> ___VARIABLE_moduleName___.FormatterOutput {
```
Comment (fgulan) [inline Templates/Resources/Templates/___FILEBASENAME___Formatter.swift.erb]: ForlatterInput -> FormatterInput

Comment (fgulan) [inline Templates/Resources/Templates/___FILEBASENAME___Formatter.swift.erb]: Add blank line before method name and after method

File: Templates/Resources/TemplateInfo.plist.erb [hunk 4]
```diff
@@ -34,17 +34,39 @@
...
+				<%- @interface.types.each do |type| -%>
+				<string><%= type.name %></string>
+				<%- end -%>
+			</array>
+		</dict>
+		<dict>
+			<key>Identifier</key>
+			<string>complex</string>
+			<key>Name</key>
+			<string>Module Complexity:</string>
```
[the remark this answers is not in the document]
Comment (Truba) [reply Templates/Resources/TemplateInfo.plist.erb]:
```
Simple
With Interactor
With Interactor & Formatter
```
?

File: Templates/Resources/TemplateInfo.plist.erb [hunk 5]
```diff
@@ -34,17 +34,39 @@
...
 		<dict>
 			<key>Identifier</key>
-			<string>Storyboard</string>
+			<string>interface</string>
 			<key>Name</key>
-			<string>Generate Storyboard with UIViewController</string>
+			<string>Generate Interface Builder:</string>
 			<key>Description</key>
-			<string>Generate Storyboard with UIViewController</string>
+			<string>Whicih interface builder to generate</string>
```
[the remark this answers is not in the document]
Comment (Truba) [reply Templates/Resources/TemplateInfo.plist.erb]: 👍

File: Templates/Resources/TemplateInfo.plist.erb [hunk 6]
```diff
@@ -34,17 +34,39 @@
...
+				<%- end -%>
+			</array>
+		</dict>
+		<dict>
+			<key>Identifier</key>
+			<string>complex</string>
+			<key>Name</key>
+			<string>Module Complexity:</string>
+			<key>Description</key>
+			<string>Estimated complexity of modul to be generated</string>
```
[the remark this answers is not in the document]
Comment (Truba) [reply Templates/Resources/TemplateInfo.plist.erb]: 👍

File: Templates/Resources/Templates/___FILEBASENAME___Interfaces.swift.erb [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    <%- if @complexity.generate_formatter -%>
+    struct FormatterInput {
+    }
+
+    struct FormatterOutput {
+    }
+    <%- end -%>
+
```
[the remark this answers is not in the document]
Comment (Truba) [reply Templates/Resources/Templates/___FILEBASENAME___Interfaces.swift.erb]: 👍

File: Templates/Resources/Templates/___FILEBASENAME___Presenter.swift.erb [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+import RxCocoa
+<% end -%>
+
+final class ___VARIABLE_moduleName___Presenter {
+
+    // MARK: - Private properties -
+
+    private unowned var view: ___VARIABLE_moduleName___ViewInterface
+<% if @complexity.generate_formatter -%>
+    private var formatter: ___VARIABLE_moduleName___FormatterInterface
```
[the remark this answers is not in the document]
Comment (Truba) [reply Templates/Resources/Templates/___FILEBASENAME___Presenter.swift.erb]: I forgot about let, but community mostly does't use underscores.

File: Templates/Resources/Templates/___FILEBASENAME___Interfaces.swift.erb (hunk 1 above)
Comment (Truba) [reply Templates/Resources/Templates/___FILEBASENAME___Interfaces.swift.erb]: 👍

File: Templates/Resources/Templates/___FILEBASENAME___ViewController.swift.erb (hunk 2 above)
Comment (Truba) [reply Templates/Resources/Templates/___FILEBASENAME___ViewController.swift.erb]: Community mostly does't use underscores.

File: Templates/Resources/TemplateInfo.plist.erb (hunk 4 above)
[the remark this answers is not in the document]
Comment (fgulan) [reply Templates/Resources/TemplateInfo.plist.erb]: 👍

**Pull request merged**